lib/cext/preprocess.rb

```diff
@@ -115,11 +115,24 @@ def self.patch(file, contents, directory)
   end
 
   if __FILE__ == $0
-    puts "#line 1 \"#{ARGF.filename}\""
+    require 'stringio'
+    output_io = StringIO.new
+
+    file_name = ARGF.filename
+    original_content = File.read(file_name)
+    content = patch(file_name, original_content.clone, Dir.pwd)
+    content.each_line do |line|
+      output_io.puts preprocess(line)
+    end
+
+    output = output_io.string
+    $stdout.puts "#line 1 \"#{file_name}\""
+    $stdout.puts output
 
-    contents = patch(ARGF.filename, File.read(ARGF.filename), Dir.pwd)
-    contents.each_line do |line|
-      puts preprocess(line)
+    if ENV['PREPROCESS_DEBUG'] && original_content != output
+      patched_file_name = ".#{File.basename file_name, '.*'}.patched#{File.extname file_name}"
```

Comment:
Should it be output in `File.dirname`, as a sibling file of the original?
I think that would be convenient. In that case, it's missing a `#{File.dirname(file_name)}/` prefix.

Reply:
`Pathname#sub_ext` could be convenient and shorter here.

Reply:
It already is a sibling of the original. sub_ext does not allow to reuse the actual extension.

Reply:
What if `file_name` is something like `"foo/bar.c"`? Or can this not happen?
If that can happen, then the subdirectory `foo` is lost in `patched_file_name`.

Reply:
File.basename removes the subdirectory:
```ruby
> File.basename("foo/bar.c", ".*")
=> "bar"
```

That's why I suggested sub_ext, which doesn't do that:
```ruby
Pathname("foo/bar.c").sub_ext(".patched.c")
# => foo/bar.patched.c
```
But that doesn't allow to add the `.` in front of the basename, which I guess is what you meant.

Reply:
ah, yeah that might be a problem. So far all patched files are not in any subdir so it did not manifest.
I meant not to make an assumption that the suffix is .c it could also be .cpp I think.